File: src/main/java/insogroup/utils/utils/IOC/strategies/RedirectIOCStrategy.java
package insogroup.utils.utils.IOC.strategies;

import insogroup.utils.utils.IOC.IOC;
import insogroup.utils.utils.IOC.exceptions.ResolveStrategyException;
import insogroup.utils.utils.IOC.interfaces.IOCStrategy;

/**
 * Simple realisation of redirect IOCStrategy
 */
public class RedirectIOCStrategy implements IOCStrategy {
    protected String targetStrategy;

    /**
     * @param targetStrategy Strategy which will used for resolving dependencies
     */
    public RedirectIOCStrategy(String targetStrategy) {
        this.targetStrategy = targetStrategy;
    }

    @Override
    public Object getInstance(Object... args) throws ResolveStrategyException {
        return IOC.resolve(targetStrategy, args);
    }
}

File: src/main/kotlin/com/github/insanusmokrassar/utils/IOC/strategies/RedirectIOCStrategy.kt
package com.github.insanusmokrassar.utils.IOC.strategies

import com.github.insanusmokrassar.utils.IOC.exceptions.ResolveStrategyException
import com.github.insanusmokrassar.utils.IOC.getOrCreateIOCInstance
import com.github.insanusmokrassar.utils.IOC.interfaces.IOCStrategy

/**
 * Simple realisation of redirect IOCStrategy
 * @param targetStrategy Strategy which will used for resolving dependencies
 */
class RedirectIOCStrategy(iocName: String, private val targetStrategy: String) : IOCStrategy {

    private val ioc = getOrCreateIOCInstance(iocName)

    @Throws(ResolveStrategyException::class)
    override fun <T> getInstance(vararg args: Any): T {
        return ioc.resolve<T>(targetStrategy, *args)
    }
}
